Add functional commitment demo scripts.

porcuquine opened the pull request:
This PR adds some simple scripts demonstrating functional commitments and chained functional commitments. It's probably useful to walk through them to understand the mechanism being demonstrated. They are fairly simple and clear, and the proofs don't take long to generate.

I've also added several proofs to the final chained-functional-commitment example in `demo/alpha.lurk`. In that demo, we fairly rapidly build up a simple relational-ish database and use it to implement a ledger that disallows illegal balance transfer attempts.

The final chained applications and proofs demonstrate the use of functional commitments to the head of sequential blockchain states. It's obviously a simplified example, but it moves pretty quickly from the base language to a 'main event' that should make sense to many.

If you want to help create these proofs, please comment in response to the comments I will put in the source. If you succeed, you'll need to harvest the proof out of `/var/tmp/fcomm_data/nova_proofs`. You'll see the file name in the output of `(:prove)`. That's the file to capture. Make sure to grab it right away, since it might get deleted if you restart your machine…

You can also put something like this in your shell init file:
```
export FCOMM_DATA_PATH=$HOME/.clutch/data
```

porcuquine commented on clutch/demo/alpha.lurk:
```diff
@@ -121,7 +124,36 @@
+;; ((:ID 0 :BALANCE 122 :LAST-NAME "Church" :FIRST-NAME "Alonzo") (:FIRST-NAME "Alan" :LAST-NAME "Turing" :BALANCE 456 :ID 1) (:ID 2 :BALANCE 9001 :LAST-NAME "Nakamoto" :FIRST-NAME "Satoshi"))
+
+
+; !(:prove)
```
This one needs a proof. To make it, edit the source to uncomment the line (by removing the initial `;`), then step through the demo (by hitting return) until you get to this line. Then hit return one more time and wait for it to return. It could take a long time.

porcuquine commented on clutch/demo/alpha.lurk:
```diff
@@ -121,7 +124,36 @@
+;; (comm 0x363b52b6be70c524df767e3a0f0289e5a1ebb0abf2c04d01fe4078b18627a66c)
+;; ((:FIRST-NAME "Alonzo" :LAST-NAME "Church" :BALANCE 117 :ID 0) (:FIRST-NAME "Alan" :LAST-NAME "Turing" :BALANCE 456 :ID 1) (:FIRST-NAME "Satoshi" :LAST-NAME "Nakamoto" :BALANCE 9006 :ID 2))
+
+; !(:prove)
```
This one needs a volunteer.

porcuquine commented on clutch/demo/alpha.lurk:
```diff
@@ -121,7 +124,36 @@
+;; (comm 0x1f5fdf82f21b6c6887de6dccc2423b6ce176cf0a1186e6ab49401ca3cd96d45d)
+;; ((:ID 0 :BALANCE 137 :LAST-NAME "Church" :FIRST-NAME "Alonzo") (:ID 1 :BALANCE 436 :LAST-NAME "Turing" :FIRST-NAME "Alan") (:FIRST-NAME "Satoshi" :LAST-NAME "Nakamoto" :BALANCE 9006 :ID 2))
+
+; !(:prove)
```
This one needs a volunteer.

porcuquine replied:
I, @porcuquine, am running this one now.

huitseeker commented on clutch/demo/alpha.lurk:
```diff
@@ -121,7 +124,36 @@
 
 !(:call 0x36283c9973a837574e3e5a15815b60a541df265b0f218ab048e99a7c44ee3769 '(1 0 2))
-;; !(:prove)
+!(:prove)
```
Won't folks need the proof for this in order to not be stuck here for 2h?
For reference, it's at 
https://gist.github.com/huitseeker/f287daaa4fab64ee14d3d5b4fd7b06ec

porcuquine replied:
Yes, good point. However, as an aside — I tried changing the reduction-count to 10, and now I can generate all of these proofs fairly quickly and without errors. I'll update the PR, then we can merge. I no longer need a swarm of helpful provers for this purpose.

huitseeker replied:
You don't, but I would bet a certain @johnchandlerburnham would. His ambitions are much more size-challenged, or so I've heard.

porcuquine replied:
I was only referring to the call-to-provers this PR itself represented — not to the general value of proving services/volunteers.

porcuquine commented:
Because I discovered that using 10 reductions per step is dramatically faster/better for me, I switched to make that the default. In order for this not to cause problems with incompatible proofs (depending on this parameter), I added support for distinct proof directories (see #303).

Because `chunk_frame_count` is a bad name, and because we have been using `reduction_count` in `fcomm`, I also changed this everywhere. Henceforth, let's always refer to 'reduction_count' in code and conversation.

porcuquine commented:
I also added support for reduction-count 100, but because the proofs become large, this makes verification (looked up by claim CID) slower (time to load the proof), so I left the clutch default as 10, at least for purposes of this demo.

Incidentally, we should work on a less verbose format for serialized proofs, since actually minimizing number of proof bytes is important for real use.

huitseeker commented on clutch/src/lib.rs:
```diff
@@ -102,7 +102,9 @@ impl<F: LurkField> ClutchState<F> {
 impl ReplTrait<F> for ClutchState<F> {
     fn new(s: &mut Store<F>, limit: usize, command: Option<Command>) -> Self {
-        let proof_map = fcomm::nova_proof_cache();
+        let reduction_count = 100;
```
It would be good to have this as a file constant (if not a constructor parameter), in order to avoid the magic number in code.

huitseeker commented on src/proof/nova.rs:
```diff
@@ -65,22 +65,22 @@ impl<'a> MultiFrame<'a, S1, IO<S1>, Witness<S1>> {
 
 pub struct NovaProver<F: LurkField> {
-    chunk_frame_count: usize,
+    reduction_count: usize,
```
A one-liner comment here saying that each Nova step will carry over `reduction_count` steps of the small-step abstract amchine of Lurk would help.

huitseeker commented on fcomm/src/lib.rs:
```diff
@@ -60,8 +61,8 @@ mod base64 {
 pub type NovaProofCache = FileMap<Cid, Proof<'static, S1>>;
-pub fn nova_proof_cache() -> NovaProofCache {
-    FileMap::<Cid, Proof<S1>>::new("nova_proofs").unwrap()
+pub fn nova_proof_cache(rc: usize) -> NovaProofCache {
```
An expanded `reduction_count` parameter name would help.

huitseeker reviewed (commented):
I highlighted a few UX issues, nothing major to fix.
